Show views' title as an overlay during scale

dkondor opened the pull request:
This is an implementation for #990 

Some notes:
 - I've made this part of the title filter plugin, since I'm unsure if it makes sense to add and extra plugin and also to share code more easily; maybe the title filter plugin could be renamed to "Scale addons" or something similar?
 - The helper classes: `cairo_text_t` and `view_overlay` could be part of some common tools. This way, other plugins can re-use them to render text and to add overlays to views.
 - The `view_overlay` transformer is not efficient in its current form, since there is no need to copy the view's texture to a new framebuffer. I wonder if there is a way to add a transformer that renders to the latest framebuffer without copying that I missed, or if it would make sense to add this functionality?
 - I could not find an easy way so far to take output scaling into account; this should probably be fixed before merging, so any suggestion on this is welcome :)
 - Also, this would benefit from better damage tracking. For the current case, it is not really a problem, since the transformers added by scale will need to damage the whole view anyway.

Fixes #990

ammen99 commented:
> * I've made this part of the title filter plugin, since I'm unsure if it makes sense to add and extra plugin and also to share code more easily; maybe the title filter plugin could be renamed to "Scale addons" or something similar?

The two functionalities (filtering based on title and showing the titles themselves) seem related. So, I think it is fine to have them in the same plugin.




> * 
> * The helper classes: `cairo_text_t` and `view_overlay` could be part of some common tools. This way, other plugins can re-use them to render text and to add overlays to views.

`cairo_text_t` seems pretty generic - so it might be a good fit for `plugins/common/...`




> * 
> * The `view_overlay` transformer is not efficient in its current form, since there is no need to copy the view's texture to a new framebuffer. I wonder if there is a way to add a transformer that renders to the latest framebuffer without copying that I missed, or if it would make sense to add this functionality?

There are two ways I can think of, which will do this, with different trade-offs:

- Add a subsurface above the main surface. However, this subsurface will be affected by the scale's transformer, so the text there would get tiny if you have a lot of open windows, though maybe that's desired? This approach has the benefit that the scale's fade-in/fade-out for hidden views will also fade-in/out the text.
- Render the text as an overlay on top of the whole screen - like the filter text is drawn currently. You'll have to iterate through all views, check their scale transformer and visibility to determine whether to draw text (and what alpha to use for fade-in/out). You can use `view->get_bounding_box()` to figure out the extents of each view on the screen.

> * I could not find an easy way so far to take output scaling into account; this should probably be fixed before merging, so any suggestion on this is welcome :)

This is usually handled like this: assume output scale 2. We want to render a 100x100 rectangle with text size 20. So, we actually create a cairo surface (and then a GL texture) of size 200x200 with text size 40. I hope this is insightful enough ;)

> * 
> * Also, this would benefit from better damage tracking. For the current case, it is not really a problem, since the transformers added by scale will need to damage the whole view anyway.

True, I don't think this is a problem at the current stage, the view transformers will consume much more resources anyway.

dkondor commented:
> > * The helper classes: `cairo_text_t` and `view_overlay` could be part of some common tools. This way, other plugins can re-use them to render text and to add overlays to views.
> 
> `cairo_text_t` seems pretty generic - so it might be a good fit for `plugins/common/...`

ok, I'm moving it to cairo-util.hpp

> > * The `view_overlay` transformer is not efficient in its current form, since there is no need to copy the view's texture to a new framebuffer. I wonder if there is a way to add a transformer that renders to the latest framebuffer without copying that I missed, or if it would make sense to add this functionality?
> 
> There are two ways I can think of, which will do this, with different trade-offs:
> 
>     * Add a subsurface above the main surface. However, this subsurface will be affected by the scale's transformer, so the text there would get tiny if you have a lot of open windows, though maybe that's desired? This approach has the benefit that the scale's fade-in/fade-out for hidden views will also fade-in/out the text.

I think it is better to have the same font size for all views for consistency and readability. It would be nice to fade together with the view, now it can look a bit weird when a view is filtered out.

>     * Render the text as an overlay on top of the whole screen - like the filter text is drawn currently. You'll have to iterate through all views, check their scale transformer and visibility to determine whether to draw text (and what alpha to use for fade-in/out). You can use `view->get_bounding_box()` to figure out the extents of each view on the screen.

This sounds a bit complicated, especially when views overlap during an animation. I'll think of this a bit more if there might be a better way.

> > * I could not find an easy way so far to take output scaling into account; this should probably be fixed before merging, so any suggestion on this is welcome :)
> 
> This is usually handled like this: assume output scale 2. We want to render a 100x100 rectangle with text size 20. So, we actually create a cairo surface (and then a GL texture) of size 200x200 with text size 40. I hope this is insightful enough ;)

Ok, I think I managed this again. My problem was that I'm not sure how to access the output scale beside looking at target framebuffer's scale.

ammen99 commented:
> My problem was that I'm not sure how to access the output scale beside looking at target framebuffer's scale.

Ah, yes, sorry - you should almost never look at `output->handle->scale`, the framebuffer scale is all a plugin should ever care about.

> This sounds a bit complicated, especially when views overlap during an animation. I'll think of this a bit more if there might be a better way.

Okay, you're right, my two ideas aren't ideal. Here's a third one: replace the scale transformer with your own, which is a subclass of view_2D. This way, scale will use the view_2D bits, but in your plugin, you'll be able to draw additional stuff in there. Whatever approach we decide on could also be useful for #897. 

For the actual implementation, I think it might be a bit difficult because scale currently just keeps an std::map for each view and stores a pointer to the transformer. In general, it seems a good idea to expose this data to any scale extensions that we might write, and this way the transformer will be overwriteable. Let me know if you have better ideas, even if they involve a bit more drastic changes to how scale is implemented.

damianatorrpm commented:
I tested it and it works great, the thing is if you have a terminal the text is unreadable,
gnome solves this by having the view title slightly below the view:
https://help.gnome.org/misc/release-notes/3.6/figures/users-activities-overview.png

dkondor commented:
> > My problem was that I'm not sure how to access the output scale beside looking at target framebuffer's scale.
> 
> Ah, yes, sorry - you should almost never look at `output->handle->scale`, the framebuffer scale is all a plugin should ever care about.
> 
> > This sounds a bit complicated, especially when views overlap during an animation. I'll think of this a bit more if there might be a better way.
> 
> Okay, you're right, my two ideas aren't ideal. Here's a third one: replace the scale transformer with your own, which is a subclass of view_2D. This way, scale will use the view_2D bits, but in your plugin, you'll be able to draw additional stuff in there. Whatever approach we decide on could also be useful for #897.
> 
> For the actual implementation, I think it might be a bit difficult because scale currently just keeps an std::map for each view and stores a pointer to the transformer. In general, it seems a good idea to expose this data to any scale extensions that we might write, and this way the transformer will be overwriteable. Let me know if you have better ideas, even if they involve a bit more drastic changes to how scale is implemented.

This sounds like a good idea. I created a simpler implementation first where scale's transformer allows adding overlays (similar to how post effects can be added to the render loop -- actually, since this is not specific to scale, so `scale_transformer` might be part of the core view transformer interface). This way, the transformer does not need to be replaced, but only overlays can be added on top of the transformed view. #897 could be implemented by adding another overlay (although I'm unsure how clicks would be handled, since those are currently implemented by scale).

What I could not figure out yet is how to increase the bounding box of the view that would allow rendering outside, as suggested by @damianatorrpm and what might be needed for #897.

ammen99 commented:
> What I could not figure out yet is how to increase the bounding box of the view that would allow rendering outside

Each transformer has a `get_bounding_box` method which calculates the view bounding box, you could just return a larger rectangle.

dkondor commented:
> > What I could not figure out yet is how to increase the bounding box of the view that would allow rendering outside
> 
> Each transformer has a `get_bounding_box` method which calculates the view bounding box, you could just return a larger rectangle.

I've tried this and seems to work. Although in this case, any box passed will be padded (it does not seem to be a problem in practice though).

What does not work is that scale does not take into account this extra space when positioning view, so the overlays can overlap with other views. I'm not sure how this can be fixed, since plugins that want to add overlays would need to inform scale of the size of the padding that will be added in advance. Of course, the easiest workaround is just to increase the spacing in the settings for scale.

JordanL2 commented:
> What does not work is that scale does not take into account this extra space when positioning view, so the overlays can overlap with other views. I'm not sure how this can be fixed, since plugins that want to add overlays would need to inform scale of the size of the padding that will be added in advance. Of course, the easiest workaround is just to increase the spacing in the settings for scale.

I've been wondering why this functionality wasn't added into the scale plugin itself, with an option to enable it, since people might not want the scale title filter functionality but still want this. Wouldn't adding it into scale also make it easier to fix this issue?

dkondor commented:
> > What does not work is that scale does not take into account this extra space when positioning view, so the overlays can overlap with other views. I'm not sure how this can be fixed, since plugins that want to add overlays would need to inform scale of the size of the padding that will be added in advance. Of course, the easiest workaround is just to increase the spacing in the settings for scale.
> 
> I've been wondering why this functionality wasn't added into the scale plugin itself, with an option to enable it, since people might not want the scale title filter functionality but still want this. Wouldn't adding it into scale also make it easier to fix this issue?

Mainly to not make the scale plugin code even more bigger and complex, but also since the text rendering code was originally with the title filter plugin (but since I moved that to the common headers, this is less of a concern). I think this is still up to discussion (and the current implementation is modular enough to easily move somewhere else). I like the idea of adding overlays from external plugin though, so I'd still hope to find a way that this can be taken into account by scale's layout.

You're right that there should be a way to have a way to display the titles without enabling the title filtering. I think it is a separate question how to present these options to the user (but there is a limitation that each plugin .so corresponds to a separate panel in WCM). In a way, showing the title logically belongs to the main properties of scale. Renaming the plugin from "title filter" to something more generic could also help.

soreau commented:
> > > What does not work is that scale does not take into account this extra space when positioning view, so the overlays can overlap with other views. I'm not sure how this can be fixed, since plugins that want to add overlays would need to inform scale of the size of the padding that will be added in advance. Of course, the easiest workaround is just to increase the spacing in the settings for scale.
> > 
> > 
> > I've been wondering why this functionality wasn't added into the scale plugin itself, with an option to enable it, since people might not want the scale title filter functionality but still want this. Wouldn't adding it into scale also make it easier to fix this issue?
> 
> Mainly to not make the scale plugin code even more bigger and complex, but also since the text rendering code was originally with the title filter plugin (but since I moved that to the common headers, this is less of a concern). I think this is still up to discussion (and the current implementation is modular enough to easily move somewhere else). I like the idea of adding overlays from external plugin though, so I'd still hope to find a way that this can be taken into account by scale's layout.

Quality should come first, above coding conveniences. The plugin can be separated into multiple source files if needed.

> 
> You're right that there should be a way to have a way to display the titles without enabling the title filtering. I think it is a separate question how to present these options to the user (but there is a limitation that each plugin .so corresponds to a separate panel in WCM).

This is not necessarily true. WCM creates a separate button for each .xml file, not each .so file. This means that you are able to have multiple .xml files for a single .so file. You can also use the options in a single .xml file from multiple plugin .cpp files. The thing that is currently bound is the enabling of the plugin via the toggle in WCM. If you create a new .xml file and there is no matching .so file, WCM will still show it. Enabling it will add it to the [core] plugins list. However core will basically ignore it. This is also how wf-shell settings work. Currently, the toggle buttons are not shown for certain 'plugins' in WCM based on hard coding which ones shouldn't have it.

I actually run into this situation recently where I am left wondering what to do when you might want multiple .xml files but only one shares the name of the .so. Perhaps we can simply add a flag to tell WCM not to add a toggle checkbox for the .xml containing the flag. This flag would be added to the list of .xml files WCM currently has hard coded to not show the checkbox.

> In a way, showing the title logically belongs to the main properties of scale. Renaming the plugin from "title filter" to something more generic could also help.

ammen99 reviewed (commented):
So far I have only tested it, and it kinda works, however, the title is not shown if I rapidly exit and start scale again.

ammen99 commented:
> Mainly to not make the scale plugin code even more bigger and complex, but also since the text rendering code was originally with the title filter plugin (but since I moved that to the common headers, this is less of a concern). I think this is still up to discussion (and the current implementation is modular enough to easily move somewhere else). I like the idea of adding overlays from external plugin though, so I'd still hope to find a way that this can be taken into account by scale's layout.

@dkondor What exactly is it that you are trying to achieve? Do you just want to increase the spacing option? We could add some more hooks, maybe a `scale-compute-spacing` signal if needed.

> You're right that there should be a way to have a way to display the titles without enabling the title filtering. I think it is a separate question how to present these options to the user (but there is a limitation that each plugin .so corresponds to a separate panel in WCM). In a way, showing the title logically belongs to the main properties of scale. Renaming the plugin from "title filter" to something more generic could also help.

AFAIU the implementation of the show-titles and the actual filtering is more or less independent? We could make one `scale-titles` plugin and one `scale-title-filter` plugin.

> I actually run into this situation recently where I am left wondering what to do when you might want multiple .xml files but only one shares the name of the .so. Perhaps we can simply add a flag to tell WCM not to add a toggle checkbox for the .xml containing the flag. This flag would be added to the list of .xml files WCM currently has hard coded to not show the checkbox.
> 

Sure, more metadata can be added, although here (in our particular case) we're speaking of 2 (maybe 3) separate .so files, each with their own XML file?

dkondor commented:
> @dkondor What exactly is it that you are trying to achieve? Do you just want to increase the spacing option? We could add some more hooks, maybe a `scale-compute-spacing` signal if needed.

Yes, exactly. Actually, when I think of it, this could be handled as part of the `scale-filter` signal, since that is always emitted before doing the new layout. But let me know if you think it is cleaner to have a separate signal for it.

dkondor commented:
> > 
> 
> This is not necessarily true. WCM creates a separate button for each .xml file, not each .so file. This means that you are able to have multiple .xml files for a single .so file. You can also use the options in a single .xml file from multiple plugin .cpp files. The thing that is currently bound is the enabling of the plugin via the toggle in WCM. If you create a new .xml file and there is no matching .so file, WCM will still show it. Enabling it will add it to the [core] plugins list. However core will basically ignore it. This is also how wf-shell settings work. Currently, the toggle buttons are not shown for certain 'plugins' in WCM based on hard coding which ones shouldn't have it.
> 
> I actually run into this situation recently where I am left wondering what to do when you might want multiple .xml files but only one shares the name of the .so. Perhaps we can simply add a flag to tell WCM not to add a toggle checkbox for the .xml containing the flag. This flag would be added to the list of .xml files WCM currently has hard coded to not show the checkbox.
> 
> >

@soreau thanks for the explanation! I was actually thinking of the opposite: having multiple .so files share just one .xml (and one button in WCM), so that the options for showing the title overlay can be on the same page as the main scale options. But in this case, there is no way to enable / disable the .so file that does not share the name with the plugin defined in the .xml, right?

I think for now I'll just go with @ammen99 's suggestion of moving the title overlay to a separate plugin and this question can be revisited later. From a UI perspective, it is a tradeoff between having more buttons in WCM or more options on each page. I think I'm very not typical in that I like to play with settings, so I cannot really judge which is better for "normal" users :)

dkondor commented:
> So far I have only tested it, and it kinda works, however, the title is not shown if I rapidly exit and start scale again.

I see; I've added a fix for this.

dkondor commented:
A new issue I just realized is that if scale-title-filter is loaded before scale, it gives dynamic linking errors since I put some of the implementations of scale_transformer in scale.cpp. I'm thinking of what is the best way to fix this:
 - Just move everything to the header (simplest; also, it is not a lot of code, so it is not really an issue)
 - Move scale_transformer to core (and rename it to something like view_2D_overlay), since it is not specific to scale anymore
 - Bigger idea: define dependence relationships between the plugins. This would allow add-on plugins to depend on their main functionality and also compiling common utilities (like the cairo-util.hpp) to separate "plugins".

soreau commented:
> @soreau thanks for the explanation! I was actually thinking of the opposite: having multiple .so files share just one .xml (and one button in WCM), so that the options for showing the title overlay can be on the same page as the main scale options. But in this case, there is no way to enable / disable the .so file that does not share the name with the plugin defined in the .xml, right?

Yes this is correct. I think if you're sharing the .xml though, why have multiple .so files at all? I would put everything in the same plugin (maybe separate source files) and have one .xml with multiple `<group><\group>` tags to separate them in different tabs within the plugin in WCM.

dkondor commented:
> > @soreau thanks for the explanation! I was actually thinking of the opposite: having multiple .so files share just one .xml (and one button in WCM), so that the options for showing the title overlay can be on the same page as the main scale options. But in this case, there is no way to enable / disable the .so file that does not share the name with the plugin defined in the .xml, right?
> 
> Yes this is correct. I think if you're sharing the .xml though, why have multiple .so files at all? I would put everything in the same plugin (maybe separate source files) and have one .xml with multiple `<group><\group>` tags to separate them in different tabs within the plugin in WCM.

I see you point; I'll try how it works this way.

ammen99 commented:
> * Just move everything to the header (simplest; also, it is not a lot of code, so it is not really an issue)

Sounds like a plan :)

> * Move scale_transformer to core (and rename it to something like view_2D_overlay), since it is not specific to scale anymore

I don't like this, it is quite specific to scale, even if it does not reference scale's structs directly.

> * Bigger idea: define dependence relationships between the plugins. This would allow add-on plugins to depend on their main functionality and also compiling common utilities (like the cairo-util.hpp) to separate "plugins".

So far, I have tried to avoid imposing order on plugin loading, and we don't have any dependency so far (except maybe glib-main-loop from w-p-e but that's something special).

> Yes this is correct. I think if you're sharing the .xml though, why have multiple .so files at all? I would put everything in the same plugin (maybe separate source files) and have one .xml with multiple `<group><\group>` tags to separate them in different tabs within the plugin in WCM.

Also sounds sensible, just try to keep the implementation separate (i.e if possible, don't call functions directly between the parts, and keep using signals as we currently do) - the idea being that in this way, we keep scale as simple as possible, and also allow creation of other scale-extension-plugins out-of-tree.

dkondor commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,296 @@
+            mouse_update_connected = false;
+        }
+    }
+},
```
I could not get uncrustify to be consistent with this formatting (all of these are part of the initializer list).

ammen99 replied:
alright, I'll push a fix when the pr is ready for merging

dkondor commented:
I've extended the `scale-filter` signal to include padding and moved the title overlays to the main plugin along with its settings. Let me know if there are any other issues.

ammen99 commented:
> I've extended the `scale-filter` signal to include padding

I don't think that padding belongs to the same signal as filtering ...

dkondor commented:
> > I've extended the `scale-filter` signal to include padding
> 
> I don't think that padding belongs to the same signal as filtering ...

OK, I'm adding it as a separate signal instead

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,190 @@
+     *                of the two values is used.
+     */
+    void add_overlay(scale_view_overlay *ol, int z_order,
+        std::unique_ptr<custom_data_t> data, const padding& pad = {0, 0, 0, 0})
```
I think padding should be called separately by the caller.

ammen99 commented on plugins/scale/scale.cpp:
```diff
@@ -1370,4 +1374,23 @@ class wayfire_scale : public wf::plugin_interface_t
 };
 
-DECLARE_WAYFIRE_PLUGIN(wayfire_scale);
+class scale_combined : public wayfire_scale
```
A cool trick ;)

I think it should be fine to just call the constructor and destructor directly from scale though. There shouldn't be any other direct calls ofc.

ammen99 commented on plugins/scale/wayfire/plugins/scale-signal.hpp:
```diff
@@ -73,4 +74,32 @@ void scale_filter_views(scale_filter_signal *signal, pred&& p)
+};
+
+/**
+ * name: scale-padding
```
Looking at it again, don't we already have the padding in the transformer? Why do we need to calculate it once more here?

dkondor replied:
The main reason for this is that calculating the layout might happen before the transformers are initialized, so scale needs to know about padding before that.
But this means that we are supplying the padding essentially twice.
Now I'm leaning toward keeping the signal, and making it scale's responsibility to add padding to the transformer. Overlays are then not allowed to extend padding (at least without triggering an update to scale that will emit this signal again). I'll add a version with this implimented.

ammen99 replied:
> The main reason for this is that calculating the layout might happen before the transformers are initialized, so scale needs to know about padding before that.

It all depends on how we implement it. We could compute the actual view size after adding the transformer (which is the moment overlays are added too).

ammen99 commented on plugins/scale/scale.cpp:
```diff
@@ -878,26 +874,34 @@ class wayfire_scale : public wf::plugin_interface_t
             {
-                double x = workarea.x + spacing + (spacing + scaled_width) * j;
-                double y = workarea.y + spacing + (spacing + scaled_height) * i;
+                double x = workarea.x + spacing +
```
Maybe I am misunderstanding something, but shouldn't `pad_left/right` be added only once (not multiplied by `j`)?

dkondor replied:
My assumption here is that each view has the same padding around it. This makes padding less flexible though.

ammen99 replied:
Wouldn't it be better if we could do this: for each view, have a fixed size available for view + padding (for example, width for each view is `workarea_width / views_in_the_row`), and then, the width of the view is the `max_width - spacing - padding`?

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+  protected:
+    wf::cairo_text_t overlay;
+    wf::cairo_text_t::params par;
+    wayfire_view view;
```
don't we have the view stored in `tr` already?

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+            (overflow && (w < std::floor(box.width * output_scale))))
+        {
+            par.output_scale = output_scale;
+            par.max_size     = {box.width, box.height};
```
isn't this basically the same as `render_overlay` ?

dkondor replied:
Yes, this can be simplified

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+    GLuint get_texture(float output_scale, wf::geometry_t& geometry)
+    {
+        wlr_box box = get_transformed_bounding_box();
+        int w = overlay.tex.width;
```
you could assign these after checking for regeneration of the overlay, then you won't have to re-assign the values.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+        int y = 0;
+        switch (pos)
+        {
+          /* TODO: Add some padding! */
```
add padding where?

dkondor replied:
Between the view and the title label (if it is above or below; in these cases, it is exactly touching it). I'm not sure if this is actually necessary, it does not look as bad as I expected.

ammen99 replied:
Alright. We can leave it for later if someone complains :D

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+        view->connect_signal("title-changed", &view_changed);
+    }
+
+    wf::signal_connection_t view_changed{[this] (auto)
```
you can use `=` to assign directly, and avoid double `{}`

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+    {
+        show_view_title_overlay = title_overlay_t::NEVER;
+        last_title_overlay = nullptr;
+        if (mouse_update_connected)
```
you can just disconnect, it is a no-op if the signal is not connected (and in this case, no need for `mouse_update_connected`)

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+    if (show_view_title_overlay == title_overlay_t::MOUSE)
+    {
+        update_title_overlay_mouse();
+        if (!mouse_update_connected)
```
you can just `mouse_update.disconnect()` and then connect.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+{
+    wayfire_view v;
+
+    /* TODO: better way to determine if scale's grab interface is active! */
```
There isn't a better way, unfortunately.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,326 @@
+        v = core.get_view_at(core.get_cursor_position());
+    }
+
+    if (v)
```
Hmm, if we have a dialog on top of a view, should we not display the title on the frontmost dialog (not on the main view)? The main view will not be visible for the most part.

damianatorrpm replied:
Modal dialogs e.g of dconf-editor commonly use empty titles.

ammen99 replied:
@damianatorrpm no, I think you missed my point. The current approach does this:

- render the base view
- render the base view's title
- render the dialog

The dialog is drawn over the title, so no title is visible at all...

dkondor replied:
I haven't really thought of dialogs, but it is true that they will occlude the parent. One solution could be to show the overlay for both parent and dialog with the exception that we don't show a dialog's title if it is empty. This is mainly a problem if the overlay is centered -- if it is on top or bottom, it will still be visible.

ammen99 reviewed (changes requested):
See individual comments.

dkondor commented on plugins/common/wayfire/plugins/common/cairo-util.hpp:
```diff
@@ -34,3 +35,263 @@ static void cairo_surface_upload_to_texture(
+     *   given in par.max_size). If it is smaller, than the result is centered along
+     *   that dimension.
+     */
+    wf::dimensions_t render_text(const char *text, const params& par)
```
No, I think string is fine

dkondor commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,190 @@
+     *                of the two values is used.
+     */
+    void add_overlay(scale_view_overlay *ol, int z_order,
+        std::unique_ptr<custom_data_t> data, const padding& pad = {0, 0, 0, 0})
```
I'm putting the stored callback into an std::shared_ptr. This way, it can store a lambda created on the fly, but also a plugin can add the same callback object to multiple transformers (if it does not need to store extra data with it). If this is too complicated, I could just use a unique_ptr, requiring a new callback for each transformer.

dkondor commented:
I've addressed most of the comments; I'm not sure what would be the best solution for dialogs and for padding (I've created a version where only the scale plugin adjusts the padding, based on the responses to the scale-padding signal). I've reorganized the commits and moved the old version here: https://github.com/dkondor/wayfire/tree/scale_show_view_title_old

soreau commented:
> I'm not sure what would be the best solution for dialogs and for padding (I've created a version where only the scale plugin adjusts the padding, based on the responses to the scale-padding signal).

Maybe child views shouldn't get a title overlay, the parent title should be enough?

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,172 @@
+     * Add a new overlay that is rendered after this transform.
+     *
+     * @param ol      The callback function that does the rendering.
+     * @param z_order Relative order; overlays are called in order.
```
The docstring needs to be updated as well :)

ammen99 commented:
> Maybe child views shouldn't get a title overlay, the parent title should be enough?

Sounds sensible to me - a single overlay on the topmost dialog (`enumerate_views().front()`) with the title of the main view.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,316 @@
+/**
+ * Class storing an overlay with a view's title, added to scale's transformer.
+ */
+class view_title_overlay
```
Forgot it on the first review, this is also missing `_t` suffix.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,172 @@
+ * with the ability to add overlays
+ */
+
+class scale_transformer;
```
also missing `_t`. Only signals don't have it.

dkondor commented:
https://github.com/WayfireWM/wayfire/pull/996#discussion_r552214215 @ammen99 you're right that scale can directly take into account the padding added to transformers and there is no need for a separate signal. I've made changes that work this way. I'm yet unsure what to do in the case when an overlay wants to update their padding (if this is needed at all).

dkondor commented:
> > Maybe child views shouldn't get a title overlay, the parent title should be enough?
> 
> Sounds sensible to me - a single overlay on the topmost dialog (`enumerate_views().front()`) with the title of the main view.

Sounds good, I'll see if I can implement this. Is there a guarantee that `enumerate_views().front()` gives the topmost dialog? (So that child views are ordered?)

ammen99 commented:
> Sounds good, I'll see if I can implement this. Is there a guarantee that `enumerate_views().front()` gives the topmost dialog? (So that child views are ordered?)

Yes, the docstring is missing that description but the list is ordered.
https://github.com/WayfireWM/wayfire/blob/master/src/view/view.cpp#L157

dkondor commented:
I've added some modifications to handle dialogs better. I've decided to add a separate callback for calculating the required padding for this (this is required to properly adjust padding when the dialog's size is changed via a relayout).

What works:
 - If the overlay position is top or bottom, then only the parent view displays an overlay
 - If the overlay position is center, only the topmost dialog displays it, adding padding if the overlay is larger than the dialog

Issues:
 - There are sometimes visual glitches during the animation when a dialog is opened / closed while scale is running (most easily reproduced with gedit's "save before closing" dialog and using the middle click close option).
 - If a scaled dialog is vertically larger than the scaled parent, it will still be on top of the overlay. I'm not sure what would be a good way to adjust the position in this case, maybe the parent can calculate the combined bounding box of its dialogs. Or simply scale can restrict dialogs to be always strictly smaller than their parents.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -403,7 +403,7 @@ void scale_show_title_t::update_title_overlay_opt()
         update_title_overlay_mouse();
         mouse_update.disconnect();
-        wf::get_core().connect_signal("pointer_motion_post", &mouse_update);
+        wf::get_core().connect_signal("pointer_motion_absolute_post", &mouse_update);
```
absolute events happen only on Wayland/X11, so this will actually break DRM backend (which is much more important)

dkondor replied:
Is it an issue to connect to both? (Although I agree that having this feature on X11 is probably not that important)

ammen99 replied:
You can connect to both, as long as the `mouse_update` handler does not care about its argument (because the `signal_data_t`s passed to the two signals have different types.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,452 @@
+
+            if (show_view_title_overlay == title_overlay_t::MOUSE)
+            {
+                auto tmp_view = view;
```
seems to me we should have a helper function `get_toplevel_view` which returns the topmost parent.

dkondor replied:
Yes, that makes sense; I've copied the one from [view.cpp](https://github.com/WayfireWM/wayfire/blob/27a5a97bec803c70b175ae6d0f84a2925c5e329a/src/view/view.cpp#L73)

dkondor commented:
I'm converting this to a draft since I'm not satisfied with how dialogs are handled. I'm now thinking that using an overlay effect would be a better solution, but it could be best if it is integrated with scale better. I'll see if I can come up with something that works that way.

ammen99 commented:
> I'm converting this to a draft since I'm not satisfied with how dialogs are handled. I'm now thinking that using an overlay effect would be a better solution, but it could be best if it is integrated with scale better. I'll see if I can come up with something that works that way.

I believe using visual tricks here would be the easiest way. We could have the titles fade-in very quickly once the scale animation is done, and then have them fade-out very quickly once scale starts exiting. It would require a pair of new signals from scale, but then you won't have to support overlays and what not in the transformer (so I think overall that solution could actually be simpler than what we have now).

dkondor commented:
> > I'm converting this to a draft since I'm not satisfied with how dialogs are handled. I'm now thinking that using an overlay effect would be a better solution, but it could be best if it is integrated with scale better. I'll see if I can come up with something that works that way.
> 
> I believe using visual tricks here would be the easiest way. We could have the titles fade-in very quickly once the scale animation is done, and then have them fade-out very quickly once scale starts exiting. It would require a pair of new signals from scale, but then you won't have to support overlays and what not in the transformer (so I think overall that solution could actually be simpler than what we have now).
> 
> EDIT: this is just one more idea I had, feel free to do something else if you have better ideas :)

The problem is that another plugin can add a transformer and run an animation and we need to adapt to that as well (e.g. if that results in the scaled size of the view changing.

I feel the better way to deal with that is to always run a pre-render hook that reacts to any changes in state + also make sure we detect any change in the view's transformed bounding box between the pre-render hook and the actual rendering.

I've put together an implementation that now works in all the corner cases that I could think of (including multiple dialogs, etc.). Let me know if you agree with this direction.

Also, since this turned out to be more complex than I originally expected, any more testing is much appreciated :)

@JordanL2 @damianatorrpm @soreau let me know if you have any more feedback on the implementation

JordanL2 commented:
> @JordanL2 @damianatorrpm @soreau let me know if you have any more feedback on the implementation

I don't think I have the experience to provide any valuable insight on the implementation but thanks very much for doing this work :)

damianatorrpm commented:
@dkondor I really like it a lot and couldn't find any issues with it.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,181 @@
+        std::shared_ptr<scale_view_overlay> ol; /* callback function */
+        int z_order; /* order in which overalys are applied */
+        padding pad; /* extra padding needed by this overlay */
+        overlay(std::shared_ptr<scale_view_overlay>& ol_, int z_order_,
```
you don't need such constructors, you could directly initialize with `{}`

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,181 @@
+        {}
+    };
+
+    std::list<std::unique_ptr<overlay>> overlays; /* list of active overlays */
```
Seems to me that overlays are movable and copyable - do we need to keep them in `unique_ptr`s ?

dkondor replied:
I was unsure and I switched to using shared_ptr everywhere. Now I feel it should be unique_ptr, since one overlay will go together with one view transformer. If the caller wants to share state / resources between overlays, they can do this by inheriting from the overlay struct and adding shared_ptr members.
Also, I'm not sure if the callbacks should be std::functions, or it might be better to use virtual functions (if users of overlays will typically inherit anyway).

ammen99 replied:
I see, I have missed this reorganization - I was reviewing commit-by-commit, so this particular comment is outdated.

In any case, shared_ptr is not semantically correct here. It implies shared ownership, however, an overlay should never remain in the transformer if the entity which added it is destroyed. So a unique_ptr/observer_ptr combo sounds better.
Plus, if you push observer-ptrs, you can make several such pointers from the same unique_ptr, thus sharing overlays between transformers, if you wanted to.

Another note: please rename `overlay` and `padding` with `_t` versions.

dkondor replied:
I'm leaning toward `unique_ptr` -- this would allow to create, add and "forget" an overlay (and anyway, in most cases, there needs to be some extra data that is unique to every view).

ammen99 replied:
Yeah, this is what we do for ex. for view transformers and custom data. There are some pitfalls though - you still need to ensure all of these are removed when the plugin is unloading, otherwise, it could happen that the .so file is unloaded (and thus the destructor cannot run anymore).

dkondor replied:
That's true. For now, the title-overlay is compiled in the same library as scale, so this will not be a problem, but for other uses, this means there is a need to keep track of overlays added (but I'm not sure if using `observer_ptr` or `shared_ptr` would solve it).

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    /* remove an existing overlay */
+    void rem_overlay(scale_view_overlay *ol)
```
IMO this is inconsistent, if we pass shared ptrs for adding, we should use shared ptrs for removing too. (or unique_ptr + observer_ptrs, but maybe shared_ptrs are easier here)

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,309 @@
+     */
+    wlr_box get_transformed_bounding_box()
+    {
+        wlr_box box = tr.get_transformed_view()->get_wm_geometry();
```
You should use the `get_bounding_box(nonstd::observer_ptr<transformer>)` functions, they will give you the bounding box up to the transformer, and then you can apply the last step manually. `wm_geometry` is rarely a good option since you can have scaling, etc.

dkondor replied:
The problem is that `get_bounding_box()` includes shadows. See the attached screenshot where the gedit window was maximized, so no shadows, while the other ones were not, so there is an extra gap below the overlay. This is why I found it is better to use `get_wm_geometry()` and transform it ourselves: the `trasform_box_without_padding()` function that calls `transform_region(this)` to do the transform up to our transformer.

dkondor replied:
![scale_overlay_padding](https://user-images.githubusercontent.com/6224679/104732647-8096f180-5778-11eb-940c-37c7f4cb2c53.png)

ammen99 replied:
I see, makes sense. Can we then rename the function to something that indicates it is not the full bounding box?

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,309 @@
+        view->damage();
+    }
+
+    void render_overlay()
```
nitpick: a better name would be `update_overlay_texture` or similar. `render_overlay` sounds like you are actually rendering it on the screen.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -0,0 +1,309 @@
+
+void scale_show_title_t::fini()
+{
+    output->disconnect_signal(&view_filter);
```
no need for manually disconnecting actually, because the signal_connection_t objects will do this automatically.

ammen99 commented on plugins/scale/scale-title-overlay.cpp:
```diff
@@ -19,28 +47,90 @@ class view_title_overlay_t
+            view->get_transformer(wf::scale_transformer_t::transformer_name());
+        auto tr = dynamic_cast<wf::scale_transformer_t*>(tmp.get());
+        if (!tr)
+        {
```
you could just assert(false), since this should never happen, if it does, at least we'll be able to detect such mistakes, otherwise, we might spent hours trying to figure out what is actually going on :)

dkondor replied:
I'm not completely sure, since this is called from [here](https://github.com/dkondor/wayfire/blob/9f78ba1c08f9706ff0d04ed66db18f29e01893b1/plugins/scale/scale-title-overlay.cpp#L209) and I wonder if that line can be reached before the scale plugin added its transformer to a newly created dialog. Still, the test could be `if(!tmp)` and the `dynamic_cast` failing can be an error.

ammen99 replied:
Oh hm, this is a very good point. In this case, could you add a comment about this? i.e how we can reach that point.

dkondor replied:
Ok, sure. I'm adding comments here. I'm not entirely sure if this will happen though, but I'm not that familiar with the sequence of events when views are created.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -90,75 +100,103 @@ class scale_transformer_t : public wf::view_2D
-            return other->z_order >= z_order;
-        });
+            auto& ol = *(p.second);
+            if (ol.pre_hook)
```
Can't we just start with empty pad, extend padding by going through all hooks, then compare with last padding? Seems like it would be simpler and shorter.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -90,75 +100,103 @@ class scale_transformer_t : public wf::view_2D
     {
-        add_overlay(ol, z_order, pad);
-    }
+        if (!ol)
```
This should not be here. Callers need to ensure they provide sensible values. I'd assert() if anything.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -90,75 +100,103 @@ class scale_transformer_t : public wf::view_2D
+            return other.first >= z_order;
+        });
+
+        view_padding.extend(ol->view_padding);
```
won't padding be updated by the prehook anyway?

dkondor replied:
It would, but this way, there is no need to call `damage()` twice; also, this will work if this call happens between our pre-hooks and rendering (I'm not sure why that would happen, but it would be completely OK to do that, e.g. from another plugin's pre-hook)

ammen99 replied:
ok, sounds sensible.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -232,6 +243,21 @@ class scale_transformer_t : public wf::view_2D
      */
     wlr_box get_bounding_box(wf::geometry_t view, wlr_box region) override
     {
+        if (view != last_view_box)
```
Why is this needed? If another transformer has changed, it should call damage() (before rendering). When it calls damage(), this method will be called, so I don't think you need an idle call and special handling here (even if the pre hooks indicate a size change).

dkondor replied:
I see, so there is no need to call damage here, just update our padding

ammen99 reviewed (changes requested):
Starting to look good, I hope you won't have to reiterate on the design anymore :)

A few small requests, and then we can merge.

dkondor commented:
I've addressed these issues (see my individual comments) -- let me know if there's more

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,282 @@
+     * Overlays that can be added to this transformer. Hooks are called
+     * similarly to render-manager.
+     */
+    struct overlay_t : public wf::custom_data_t
```
why does this need to derive from custom_data_t ?

dkondor replied:
You're right, it can have its own virtual destructor; I'll change it.

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    /* call pre-render hooks; the parameter indicates if it is possible to
```
I find this a bit confusing. It both damages and returns whether damaging is necessary? Looking at the current usages, it seems we could directly damage (if damaging is possible at all).

dkondor replied:
This is a small optimization to avoid calling `damage()` twice, since this function is called from [here](https://github.com/dkondor/wayfire/blob/8d68476114371820ccb58a301bc70079476e7d23/plugins/scale/scale.cpp#L694) where damage might be necessary anyway.

I could either remove this if it is not worth the code complexity (since the runtime saving will be small), or use an alternate implementation with a second bool parameter that is set to true if the caller needs to damage the view. This would allow to only damage once even if the overlays' total size has shrunk.

dkondor replied:
I've changed the logic here, and made the version with the `can_damage` parameter protected since it is only used when this is called from `get_bounding_box()`

ammen99 commented on plugins/scale/wayfire/plugins/scale-transform.hpp:
```diff
@@ -0,0 +1,296 @@
+    uint32_t get_z_order() override
+    {
+        /* ensure that scale comes after fire */
+        return wf::TRANSFORMER_HIGHLEVEL + 2;
```
Unfortunately, this can't be. Scale needs to come before wobbly so that dragging a view works, see #1020 

Also, why should scale come after fire? Makes no sense.

ammen99 replied:
So, this should be highlevel - 10 or something like that.

dkondor replied:
Ok, that's reasonable. The motivation for putting this after fire was that if an overlay adds padding, then fire will cover the whole (now padded) area -- this can look strange (e.g. if a wide overlay is over a small dialog). But this is a general problem with fire (e.g. fire animation with a 45-degree rotated view will also look strange), so this can be a separate issue.

ammen99 reviewed (approved):
Thanks for working on this and thanks for the patience during the reviews!

Looks good to me, and I also tested it, only the Z order is wrong. Once you update it, please also squash your commits together, maybe like this:

- First commit, extracts the text helper
- Other changes to scale-title-filter
- Add support to scale for overlays
- Add title overlay

dkondor commented:
OK, I've made the remaining changes; let me know if there is something else
